src/reader_async.rs

```diff
@@ -160,34 +131,83 @@ impl AsyncParquetFile {
+    ///
+    ///    - `batchSize`: The number of rows in each batch. If not provided, the upstream parquet
+    ///           default is 1024.
+    ///    - `rowGroups`: Only read data from the provided row group indexes.
+    ///    - `limit`: Provide a limit to the number of rows to be read.
+    ///    - `offset`: Provide an offset to skip over the given number of rows.
+    ///    - `columns`: The column names from the file to read.
+    #[wasm_bindgen]
+    pub async fn read(&self, options: Option<ReaderOptions>) -> WasmResult<Table> {
+        let options = options
+            .map(|x| x.try_into())
+            .transpose()?
+            .unwrap_or_default();
+        let builder = create_builder(&self.reader, &self.meta, &options)?;
+
+        let schema = builder.schema().clone();
+        let stream = builder.build()?;
+        let batches = stream.try_collect::<Vec<_>>().await.unwrap();
+
+        Ok(Table::new(schema, batches))
     }
 
+    /// Create a readable stream of record batches.
+    ///
+    /// Each item in the stream will be a {@linkcode RecordBatch}.
+    ///
+    /// @param options
+    ///
+    ///    Options for reading Parquet data. Optional keys include:
+    ///
+    ///    - `batchSize`: The number of rows in each batch. If not provided, the upstream parquet
+    ///           default is 1024.
+    ///    - ~~`rowGroups`: Only read data from the provided row group indexes.~~ **The rowGroups parameter does not yet take effect here.**
+    ///    - `limit`: Provide a limit to the number of rows to be read.
+    ///    - `offset`: Provide an offset to skip over the given number of rows.
+    ///    - `columns`: The column names from the file to read.
+    ///    - `concurrency`: The number of concurrent requests to make
     #[wasm_bindgen]
     pub async fn stream(
         &self,
-        concurrency: Option<usize>,
+        options: Option<ReaderOptions>,
     ) -> WasmResult<wasm_streams::readable::sys::ReadableStream> {
-        self.inner_stream(
-            concurrency,
-            &self.meta,
-            &self.reader,
-            &self.batch_size,
-            &self.projection_mask,
-        )
-        .await
+        let options: JsReaderOptions = options
+            .map(|x| x.try_into())
+            .transpose()?
+            .unwrap_or_default();
+
+        let concurrency = options.concurrency.unwrap_or_default().max(1);
+
+        // TODO: enable row group selection
+        // let row_groups = options
+        //     .row_groups
+        //     .clone()
+        //     .unwrap_or_else(|| (0..self.meta.metadata().num_row_groups()).collect());
+        // let row_groups_slice = row_groups.as_slice();
+
+        let num_row_groups = self.meta.metadata().num_row_groups();
+        let reader = self.reader.clone();
+        let meta = self.meta.clone();
+
+        let buffered_stream = stream::iter((0..num_row_groups).map(move |i| {
```

Comment:
```diff
- let buffered_stream = stream::iter((0..num_row_groups).map(move |i| {
+ let buffered_stream = stream::iter(row_groups.into_iter().map(move |i| {
```

Reply:
oh was this just an issue with me using `iter()` instead of `into_iter()`? 🤦‍♂️

Reply:
Eh, that was me moments after saying "this thing <gestures at very similar Py03 dyn downcasting> is clearly impossible in JS" 🫠